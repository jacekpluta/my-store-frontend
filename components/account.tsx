import Head from "next/head";
import Link from "next/link";
import React, { useState } from "react";
import { Icon } from "semantic-ui-react";
import {
  AccountPageStyles,
  AccountPageTitle,
} from "./styles/AccountPageStyles";

function Account() {
  const [activeTab, setActiveTab] = useState("details");

  return (
    <>
      <Head>My Account</Head>
      <AccountPageTitle>My Account</AccountPageTitle>
      <AccountPageStyles activeTab={activeTab}>
        <div className="menu">
          <ul>
            <li>
              <Icon style={{ marginRight: "10px" }} name="user" />
              My details
            </li>
            <li>
              <Link
                href={{
                  pathname: "/orders",
                }}
              >
                <>
                  <Icon style={{ marginRight: "10px" }} name="ordered list" />
                  My orders
                </>
              </Link>
            </li>
            <li>
              <Icon style={{ marginRight: "10px" }} name="setting" />
              Account settings
            </li>
          </ul>
        </div>

        {activeTab === "details" && (
          <div className="details">
            <h2>My Details</h2>
          </div>
        )}

        {activeTab === "orders" && (
          <div className="orders">
            <h2>Orders</h2>
          </div>
        )}

        {activeTab === "settings" && (
          <div className="settings">
            <h2>Account Seeting</h2>
          </div>
        )}
      </AccountPageStyles>
    </>
  );
}

export default Account;
