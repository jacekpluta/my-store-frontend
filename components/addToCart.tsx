import React from "react";
import { useMutation } from "@apollo/react-hooks";
import gql from "graphql-tag";
import { CURRENT_USER_QUERY } from "./queries";

export const ADD_TO_CART_MUTATION = gql`
  mutation ADD_TO_CART_MUTATION($id: ID!) {
    addToCart(id: $id) {
      id
      quantity
    }
  }
`;
type itemId = {
  itemId: string;
};

export default function addToCart(props: itemId) {
  const [addToCart, addToCartMutation] = useMutation(ADD_TO_CART_MUTATION, {
    refetchQueries: [{ query: CURRENT_USER_QUERY }],
    awaitRefetchQueries: true,
  });

  const { itemId } = props;

  return (
    <button
      disabled={addToCartMutation.loading}
      onClick={async () => {
        await addToCart({
          variables: {
            id: itemId,
          },
        });
      }}
    >
      Add{addToCartMutation.loading ? "ing" : ""} to cart
    </button>
  );
}
