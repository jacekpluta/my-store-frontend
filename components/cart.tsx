import React from "react";
import CartStyles from "../styles/CartStyles";
import Supreme from "../styles/Supreme";
import CloseButton from "../styles/CloseButton";
import ButtonStyle from "../styles/ButtonStyles";
import { useQuery, useMutation } from "@apollo/react-hooks";
import gql from "graphql-tag";
import Error from "./ErrorMessage";
import { CURRENT_USER_QUERY } from "./Queries";
import CartItem from "./CartItem";
import formatMoney from "./FormatMoney";
import CreditCardCheckout from "./CreditCardCheckout";

export const LOCAL_STATE_QUERY = gql`
  query {
    cartOpen @client
  }
`;

export const TOGGLE_CART_MUTATION = gql`
  mutation {
    toggleCart @client
  }
`;

export default function Cart() {
  const queryData = useQuery(LOCAL_STATE_QUERY);
  const [toggleCart, toggleCartMutation] = useMutation(TOGGLE_CART_MUTATION);

  const currentUserQuery = useQuery(CURRENT_USER_QUERY);

  if (
    queryData.loading ||
    toggleCartMutation.loading ||
    currentUserQuery.loading
  )
    return <p>Loading...</p>;
  if (queryData.error || toggleCartMutation.error || currentUserQuery.error)
    return <Error error={toggleCartMutation.error || queryData.error}></Error>;

  const cartOpen = queryData?.data?.cartOpen;
  console.log(currentUserQuery);
  const user = currentUserQuery?.data?.user;

  const totalPrice = user?.cart
    ? user.cart.reduce((all, cartItem) => {
        if (cartItem.item) return all + cartItem.quantity * cartItem.item.price;
        else return;
      }, 0)
    : "";

  if (user) {
    return (
      <CartStyles open={cartOpen}>
        <header>
          <CloseButton onClick={() => toggleCart()} title="close">
            &times;
          </CloseButton>
          <Supreme>{user.name} - Cart</Supreme>
          <p>
            You have {user.cart.length} item {user.cart.length === 1 ? "" : "s"}{" "}
            in your cart
          </p>
        </header>
        <ul>
          {user.cart.map((cartItem) => (
            <CartItem key={cartItem.id} cartItem={cartItem}></CartItem>
          ))}
        </ul>
        <footer>
          <p>{formatMoney(totalPrice)}</p>
          {user.cart.length && (
            <CreditCardCheckout
              cart={user.cart}
              totalPrice={totalPrice}
              allItemsCount={user.cart.length}
              user={user}
            >
              <ButtonStyle
                onClick={() => {
                  toggleCart();
                }}
              >
                Checkout
              </ButtonStyle>
            </CreditCardCheckout>
          )}
        </footer>
      </CartStyles>
    );
  } else {
    return "You are not logged in";
  }
}
