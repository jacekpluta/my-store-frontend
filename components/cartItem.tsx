import React, { useEffect } from "react";
import styled from "styled-components";
import formatMoney from "./FormatMoney";
import { useQuery, useMutation } from "@apollo/react-hooks";
import gql from "graphql-tag";
import { CURRENT_USER_QUERY } from "./Queries";

const ButtonStyle = styled.button`
font-size: 3rem;
background: none;
border: 0;
&:hover {
    color: ${(props) => props.theme.red}
    cursor: pointer
}`;

const CartItemStyles = styled.li`
  padding: 1rem 0;
  border-bottom: 1px solid ${(props) => props.theme.lightgray};
  display: grid;
  align-items: center;
  grid-template-columns: auto 1fr auto;
  img {
    margin-right: 10px;
  }

  p {
    margin: 0;
  }
`;

export const DELETE_CART_ITEM_WHEN_ITEM_DELETED_MUTATION = gql`
  mutation DELETE_CART_ITEM_WHEN_ITEM_DELETED_MUTATION($id: ID!) {
    deleteCartItemWhenItemDeleted(id: $id) {
      id
    }
  }
`;

export const DELETE_CART_ITEM_MUTATION = gql`
  mutation DELETE_CART_ITEM_MUTATION($id: ID!) {
    deleteCartItem(id: $id) {
      id
    }
  }
`;

const GET_DELETED_ITEM_ID = gql`
  {
    id @client
  }
`;

export default function CartItem({ cartItem }) {
  const [
    deleteCartItemWhenItemDeleted,
    deleteCartItemWhenItemDeletedMutation,
  ] = useMutation(DELETE_CART_ITEM_WHEN_ITEM_DELETED_MUTATION, {
    refetchQueries: [{ query: CURRENT_USER_QUERY }],
    awaitRefetchQueries: true,
  });

  const [deleteCartItem, deleteCartItemMutation] = useMutation(
    DELETE_CART_ITEM_MUTATION,
    {
      refetchQueries: [{ query: CURRENT_USER_QUERY }],
      awaitRefetchQueries: true,
    }
  );

  const currentUserQuery = useQuery(CURRENT_USER_QUERY);
  const getDeletedItemQuery = useQuery(GET_DELETED_ITEM_ID);
  const deletedItemId = getDeletedItemQuery?.data?.id;

  const deleteCartUpdate = (cache, payload) => {
    const data = cache.readQuery({ query: CURRENT_USER_QUERY });

    const cartItemId = payload.data.deleteCartItem.id;

    const newData = data.user.cart.filter((cartItem) => {
      return cartItem.id !== cartItemId;
    });
    cache.writeQuery({ query: CURRENT_USER_QUERY, data: newData });
  };

  useEffect(() => {
    if (deletedItemId) {
      async function deleteItemFromCart() {
        const userCart = currentUserQuery.data.user.cart;

        //check if deleted item is in the cart
        const foundItem = userCart.filter((item) => {
          if (item.item.id === deletedItemId) return item;
        });

        //if it is then delete it
        if (foundItem) {
          const id = foundItem[0].id;

          await deleteCartItemWhenItemDeleted({
            variables: {
              id: id,
            },
          }).catch((err) => alert(err.message));

          //reset local query
          getDeletedItemQuery.client.writeData({ data: { id: null } });
        }
      }
      deleteItemFromCart();
    }
  }, [deletedItemId]);

  return (
    <CartItemStyles>
      <img
        width="200"
        src={cartItem.item.image}
        alt={cartItem.item.title}
      ></img>
      <div>
        <h3>{cartItem.item.title}</h3>
        <p>
          {formatMoney(cartItem.item.price * cartItem.quantity)} -{" "}
          {cartItem.quantity} &times; {formatMoney(cartItem.item.price)} each
        </p>
      </div>

      <ButtonStyle
        title="Delete Item"
        onClick={async () => {
          await deleteCartItem({
            variables: {
              id: cartItem.item.id,
            },
            update: deleteCartUpdate,
            optimisticResponse: {
              __typename: "Mutation",
              deleteCartItem: {
                __typename: "CartItem",
                id: cartItem.item.id,
              },
            },
          }).catch((err) => alert(err.message));
        }}
      >
        &times;
      </ButtonStyle>
    </CartItemStyles>
  );
}
