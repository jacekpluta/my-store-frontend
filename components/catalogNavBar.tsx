import React from "react";
import { CatalogNavBarStyles } from "./styles/CatalogNavBarStyles";
import Checkbox from "./Checkbox";
import {
  ButtonCatalogNavFilter,
  ButtonCatalogNavClear,
} from "./styles/ButtonStyles";

function CatalogNavBar() {
  return (
    <CatalogNavBarStyles>
      <h2>Shop By Brand</h2>
      <Checkbox>Nike</Checkbox>
      <Checkbox>Puma</Checkbox>
      <Checkbox>Rebook</Checkbox>
      <Checkbox>dsad</Checkbox>

      <h2>Shop By Price</h2>
      <Checkbox>$0 - $25</Checkbox>
      <Checkbox>$25 - $50</Checkbox>
      <Checkbox>$50 - $75</Checkbox>
      <Checkbox>$75 - $100</Checkbox>

      <h2>Gender</h2>
      <Checkbox>Men</Checkbox>
      <Checkbox>Women</Checkbox>

      <h2>Categories</h2>
      <Checkbox>Shoes</Checkbox>
      <Checkbox>Boots</Checkbox>
      <Checkbox>Trainers</Checkbox>
      <Checkbox>Sandals</Checkbox>
      <Checkbox>Flip Flops</Checkbox>

      <ButtonCatalogNavFilter>FILTER</ButtonCatalogNavFilter>
      <ButtonCatalogNavClear>CLEAR</ButtonCatalogNavClear>
    </CatalogNavBarStyles>
  );
}

export default CatalogNavBar;
