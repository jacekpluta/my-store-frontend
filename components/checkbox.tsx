import { useQuery } from "@apollo/react-hooks";
import React, { useEffect, useState } from "react";
import { CLEAR_FILTERS_QUERY } from "../lib/queries";
import { clearFilters } from "../lib/vars";

interface MyProps {
  changeFilter: Function;
}

const Checkbox: React.FunctionComponent<MyProps> = (props) => {
  const { children, changeFilter } = props;
  const [checked, setChecked] = useState<boolean>(false);

  const clearFiltersData = useQuery(CLEAR_FILTERS_QUERY, {
    fetchPolicy: "network-only",
  });

  useEffect(() => {
    if (!clearFiltersData.loading && !clearFiltersData.error) {
      const clearIt = clearFiltersData.data.clearFilters;

      if (clearIt) {
        setChecked(false);
        clearFilters(false);
      }
    }
  }, [checked, clearFiltersData.data]);

  return (
    <div className="checkbox-container">
      <label className="checkbox-label">
        <label style={{ paddingLeft: "30px" }}>{children}</label>
        <input
          type="checkbox"
          checked={checked}
          onChange={() => {
            setChecked(!checked);
            changeFilter(children);
          }}
        />
        <span className="checkbox-custom rectangular" />
      </label>
    </div>
  );
};

export default Checkbox;
