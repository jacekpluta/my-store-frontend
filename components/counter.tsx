import React from "react";
import {
  ButtonCounterFirst,
  ButtonCounterNumber,
  ButtonCounterSecond,
} from "./styles/ButtonStyles";
import { CounterStyles } from "./styles/SingleItemStyle";

interface ICounter {
  addCounter?: Function;
  substractCounter?: Function;
  counter?: Number;
}
const Counter = ({ addCounter, substractCounter, counter }: ICounter) => {
  return (
    <CounterStyles>
      <ButtonCounterFirst
        onClick={() => {
          if (counter < 9) {
            addCounter(counter);
          }
        }}
      >
        +
      </ButtonCounterFirst>
      <ButtonCounterNumber>{counter}</ButtonCounterNumber>
      <ButtonCounterSecond
        onClick={() => {
          if (counter > 1) {
            substractCounter(counter);
          }
        }}
      >
        -
      </ButtonCounterSecond>
    </CounterStyles>
  );
};
export default Counter;
