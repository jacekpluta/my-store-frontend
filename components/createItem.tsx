import React, { useState } from "react";
import { useMutation } from "@apollo/react-hooks";
import Form from "./styles/Form";
import gql from "graphql-tag";
import { useFormFields } from "../lib/useFormFields";
import Error from "./errorMessage";
import Router from "next/router";

export interface CreateItemProps {}

export interface CreateItemState {
  description: string;
  id: string;
  image: string;
  largeImage: string;
  price: number;
  title: string;
}

export const CREATE_ITEM_MUTATION = gql`
  mutation CREATE_ITEM_MUTATION(
    $title: String!
    $description: String!
    $price: Int!
    $image: String
    $largeImage: String
  ) {
    createItem(
      title: $title
      description: $description
      price: $price
      image: $image
      largeImage: $largeImage
    ) {
      id
    }
  }
`;

const CreateItem = () => {
  const [fields, handleFieldChange] = useFormFields({
    description: "",
    price: "",
    title: "",
  });

  const [image, setImage] = useState("");
  const [largeImage, setLargeImage] = useState("");

  interface HTMLInputEvent {
    target: HTMLInputElement & EventTarget;
  }

  const uploadFile = async (e: HTMLInputEvent) => {
    const files = e.target.files;
    const data = new FormData();
    data.append("file", files[0]);
    data.append("upload_preset", "myshop");

    const res = await fetch(
      "https://api.cloudinary.com/v1_1/dr7nwbqbz/image/upload",
      { method: "POST", body: data }
    );

    const file = await res.json();

    setImage(file.secure_url);
    setLargeImage(file.eager[0].secure_url);

    {
      image && <img src={image} alt="Upload preview" />;
    }
  };

  const [createItem, createItemtMutation] = useMutation(CREATE_ITEM_MUTATION);

  return (
    <Form
      data-test="form"
      onSubmit={async (e) => {
        e.preventDefault();

        const res = await createItem({
          variables: {
            ...fields,
            image: image,
            largeImage: largeImage,
          },
        });
        Router.push({
          pathname: "/item",
          query: { id: res.data.createItem.id },
        });
      }}
    >
      <Error error={createItemtMutation.error}></Error>
      <fieldset
        disabled={
          createItemtMutation.loading ? createItemtMutation.loading : false
        }
        aria-busy={createItemtMutation.loading}
      >
        <label htmlFor="file">
          Image
          <input
            type="file"
            id="file"
            name="file"
            placeholder={!image ? "Upload an image" : "Image uploaded"}
            required
            onChange={uploadFile}
          />
        </label>

        <label htmlFor="title">
          Title
          <input
            type="text"
            id="title"
            name="title"
            placeholder="Title"
            required
            value={fields.title}
            onChange={handleFieldChange}
          />
        </label>
        <label htmlFor="price">
          Price
          <input
            type="number"
            id="price"
            name="price"
            placeholder="Price"
            required
            value={fields.price}
            onChange={handleFieldChange}
          />
        </label>
        <label htmlFor="description">
          Description
          <input
            type="text"
            id="description"
            name="description"
            placeholder="Enter a description"
            required
            value={fields.description}
            onChange={handleFieldChange}
          />
        </label>
        <button type="submit">Submit</button>
      </fieldset>
    </Form>
  );
};

export default CreateItem;
