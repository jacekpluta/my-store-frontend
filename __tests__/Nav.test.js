import { mount } from "enzyme";
import Nav from "../components/Nav";
import { CURRENT_USER_QUERY } from "../components/Queries";
import { MockedProvider } from "@apollo/react-testing";

const fakeUser = {
  __typename: "User",
  id: "4234",
  name: "test",
  email: "test@o2.pl",
  permissions: ["ADMIN"],
  cart: [],
};

const fakeItem = {
  __typename: "Item",
  id: "abc123",
  price: 5000,
  user: null,
  image: "dog-small.jpg",
  title: "dogs are best",
  description: "dogs",
  largeImage: "dog.jpg",
};

const fakeCartItem = {
  __typename: "CartItem",
  id: "omg123",
  quantity: 3,
  item: fakeItem,
  user: fakeUser,
};

const notSignedInMocks = [
  {
    request: { query: CURRENT_USER_QUERY },
    result: { data: { user: null } },
  },
];

const signedInMocks = [
  {
    request: { query: CURRENT_USER_QUERY },
    result: { data: { user: fakeUser } },
  },
];

const signedInMocksWithCartItems = [
  {
    request: { query: CURRENT_USER_QUERY },
    result: {
      data: {
        user: { ...fakeUser, cart: [fakeCartItem, fakeCartItem, fakeCartItem] },
      },
    },
  },
];

it("renders user nav component when signed in", async () => {
  const wrapper = mount(
    <MockedProvider mocks={signedInMocks}>
      <Nav />
    </MockedProvider>
  );
  await new Promise((resolve) => setTimeout(resolve, 0));
  wrapper.update();

  const nav = wrapper.find('ul[data-test="nav"]');

  expect(nav.children().length).toBe(7);

  expect(nav.text()).toContain("Logout");
  expect(nav.text()).toContain("Items");
  expect(nav.text()).toContain("Hello");
  expect(nav.text()).toContain("Sell");
  expect(nav.text()).toContain("Orders");
  expect(nav.text()).toContain("My Cart");
});

it("renders different nav component when not signed in", async () => {
  const wrapper = mount(
    <MockedProvider mocks={notSignedInMocks}>
      <Nav />
    </MockedProvider>
  );
  await new Promise((resolve) => setTimeout(resolve, 0));
  wrapper.update();

  const nav = wrapper.find('ul[data-test="nav"]');
  expect(nav.children().length).toBe(2);
});

it("renders corrent amount of items in cart", async () => {
  const wrapper = mount(
    <MockedProvider mocks={signedInMocksWithCartItems}>
      <Nav />
    </MockedProvider>
  );
  await new Promise((resolve) => setTimeout(resolve, 0));
  wrapper.update();

  const nav = wrapper.find('ul[data-test="nav"]');
  const count = nav.find("div.count");
  expect(count).toMatchSnapshot();
});
