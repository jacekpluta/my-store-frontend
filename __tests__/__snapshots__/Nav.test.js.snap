// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders corrent amount of items in cart 1`] = `
<div
  className="Nav__Dot-zxvut8-0 fEapAn count"
>
  9
</div>
`;
